text/000-structural-and-deref.md

```diff
@@ -0,0 +1,410 @@
+    fn method(this: &Parent);
+
+    #[wasm_bindgen(extends = Parent)]
+    type Child;
+    #[wasm_bindgen(constructor)]
+    fn new() -> Child;
+    #[wasm_bindgen(method)]
+    fn method(this: &Child);
+}
+```
+
+and we could then use it like so:
+
+```rust
+#[wasm_bindgen]
+pub fn run() {
+    let parent = Parent::new();
+    parent.method();
+    let child = Child::new();
+    child.method();
+}
+```
+
+and we would today see `parent` and `child` logged to the console. Ok everything
+is working as expected so far! We know we've got `Deref<Target=Parent> for
+Child`, though, so let's say we tweak this example a bit:
+
+```rust
+#[wasm_bindgen]
+pub fn run() {
+    call_method(&Parent::new());
+    call_method(&Child::new());
+}
+
+fn call_method(object: &Parent) {
+    object.method();
+}
+```
+
+Here we'd naively (and correctly) expect `parent` and `child` to be output like
+before, but much to our surprise this actually prints out `parent` twice!
+
+The issue with this is how `#[wasm_bindgen]` treats method calls today. When you
+say:
+
+```rust
+#[wasm_bindgen(method)]
+fn method(this: &Parent);
+```
+
+then `wasm-bindgen` (the CLI tool) generates JS that looks like this:
+
+```js
+const Parent_method_target = Parent.prototype.method;
+
+export function __wasm_bindgen_Parent_method(obj) {
+    Parent_method_target.call(getObject(obj));
+}
+```
+
+Here we can see that, by default, `wasm-bindgen` is **reaching into the
+`prototype` of each class to figure out what method to call**. This in turn
+means that when `Parent::method` is called in Rust, it unconditionally uses the
+method defined on `Parent` rather than walking the protype chain (that JS
+usually does) to find the right `method` method.
+
+To improve the situation there's a `structural` attribute to wasm-bindgen to fix
+this, which when applied like so:
+
+```rust
+#[wasm_bindgen(method, structural)]
+fn method(this: &Parent);
+```
+
+means that the following JS code is generated:
+
+```js
+const Parent_method_target = function() { this.method(); };
+
+// ...
+```
+
+Here we can see that a JS function shim is generated instead of using the raw
+function value in the prototype. This, however, means that our example above
+will indeed print `parent` and then `child` because JS is using prototype
+lookups to find the `method` method.
+
+Phew! Ok with all that information, we can see that **if `structural` is omitted
+then JS class hierarchies can be subtly incorrect when methods taking parent
+classes are passed child classes which override methods**.
+
+An easy solution to this problem is to simply use `structural` everywhere, so...
+let's propose that! Consequently, this RFC proposes changing `#[wasm_bindgen]`
+to act as if all bindings are labeled as `structural`. This will not be a
+breaking change because the generated bindings will still have the same behavior
+as before, they'll just handle subclassing correctly!
+
+### Adding `#[wasm_bindgen(final)]`
```

Comment:
I don't like the name `final`, because `final` means "this method *cannot* be overridden", but we're using it to mean "this method *can* be overridden, but *if* it is overridden it will have subtly wrong behavior".

Maybe `prototype` is a good alternative? Or perhaps `static_prototype` to be even more specific.

Reply:
@Pauan I think "subtly wrong behavior" is too strong here, but rather `final` indicates that this method, like `final` indicates, will not be overridden and regardless of class passed in it'll call this precise method defined for this precise class. That's "subtly wrong" when applied incorrectly but it can also be applied correctly in which case it's not "subtly wrong".

In that sense I agree with @Migi, the name `final` invokes thoughts about the right concept, namely that calling this method will always call what's being defined.

Reply:
> @Migi but I don't think people will assume that we somehow can prevent it, since we're only providing bindings.

You would be surprised, especially when we're using well-established OOP terms.

----

> @alexcrichton I think "subtly wrong behavior" is too strong here

"subtly wrong behavior" is correct. We're talking about the user expecting it to call `Bar::some_method` but instead it calls `Foo::some_method`. What would you call that if not "subtly wrong behavior"?

> but rather final indicates that this method, like final indicates, will not be overridden

That's my point: with `final` in wasm-bindgen, you *can* override the method. And this isn't hard to do either: it's possible for class hierarchies to be spread across multiple Rust crates, meaning that a downstream crate might override some methods on an upstream crate (unaware that they shouldn't do this, because of `final`).

> That's "subtly wrong" when applied incorrectly but it can also be applied correctly in which case it's not "subtly wrong".

That's why I specifically said ***if*** the method is overridden it will have subtly wrong behavior.

Reply:
@Pauan I would not call it "subtly wrong behavior" because it simply is not. You'd have to go out of your way to write down `final` which means you probably know what you're doing, so saying that you intended a subclass method I think is wrong because a manual annotation was added to say "I don't want this behavior". 

If the `final` modifier is applied blindly without understanding it then that results in the same misunderstanding from blindly applying anything without understanding it.

Also, no, you cannot override `final`. When you say `final` you're saying you're calling precisely the method indicated, no others. When calling the `final` import function you'll always call that one, no overridden function. We're not defining JS classes in Rust so we have no control over how the JS hierarchy looks in JS, but the bindings are for the Rust hierarchy and defining what's happening in Rust.

Reply:
> Also, no, you cannot override final. When you say final you're saying you're calling precisely the method indicated, no others. When calling the final import function you'll always call that one, no overridden function

I am aware of that. You are right that it will always call `Foo.prototype.someMethod`, but that's my point: that might cause subtle (and not so subtle!) problems. As an example, consider this code:

```js
class Foo {
    constructor() {
        this.someProperty = "hi";
    }

    someMethod() {
        return this.someProperty.slice(0, 0);
    }
}

class Bar extends Foo {
    constructor() {
        super();
        this.someProperty = null;
    }

    someMethod() {
        return "";
    }
}

new Bar().someMethod()                    // returns ""
Foo.prototype.someMethod.call(new Bar())  // errors
```

This is obviously contrived, but I'm sure similar situations do happen out in the wild.

In this case `Bar` extends `Foo`, so it has the same internal state (plus maybe some additional state). It can change its own state in such a way that `Foo.prototype.someMethod` is now broken/wrong (but `Bar.prototype.someMethod` is correct).

There's plenty of other situations as well: such as some code relying upon the fact that `Bar.prototype.someMethod` is called (because it updates internal state of `Bar`), so calling `Foo.prototype.someMethod` causes weird/broken behavior.

You might say "oh, but then you shouldn't use `final`", and that's correct, and easy to see in isolation, but I'm talking about a situation like this:

1. Somebody defines a `rust-foo` Rust crate which creates bindings for the `npm-foo` package.

   Since they're not overriding any methods on the classes (and perhaps due to some confusion with how `final` works in OOP languages), they mark the methods as `final`. Looking at the `rust-foo` crate in isolation, this seems okay (since no methods are being overridden):

   ```rust
   #[wasm_bindgen(module = "npm-foo")]
   extern {
       type Foo;

       #[wasm_bindgen(method, final)]
       fn some_method(this: &Foo);
   }
   ```

2. Somebody else creates a `rust-bar` Rust crate which creates bindings for the `npm-bar` package. The `npm-bar` package has some classes which extend the classes defined in `npm-foo`, and so likewise the `rust-bar` crate puts in some `extends` attributes (which point to the bindings created in `rust-foo`):

   ```rust
   #[wasm_bindgen(module = "npm-bar")]
   extern {
       #[wasm_bindgen(extends = Foo)]
       type Bar;
   }
   ```

3. The classes in `npm-bar` override some methods, but the creator of the `rust-bar` crate is unaware of this (and even if they were aware, they cannot control the `rust-foo` crate, so they can't remove the `final` attribute).

Basically, it's a situation where *in isolation* `final` seems reasonable, but it ends up being wrong in practice.

So the docs will need to be very clear about the following points:

* Exactly what `final` does (and how it differs from `final` in OOP languages).

* You almost never need `final` (it's strongly preferred to *not* use it).

* You should only use `final` if you can guarantee that ***all*** downstream consumers of this class will *never* override the method.

Reply:
I agree about those last 3 bullet points: the docs should indeed say those things clearly.

But here's the thing: To misunderstand the `final` keyword in the way you fear would happen, you basically have to not have read the docs on `final` at all. To think that `final` *prevents* overriding is a very different meaning from what `final` actually does. But since the author of the `rust-foo` binding has to explicitly write that `final` attribute, I think it's fair to assume that he would have read the docs, at least partially.